packages/js/components/src/image-gallery/image-gallery.tsx

```diff
@@ -1,33 +1,177 @@
 /**
  * External dependencies
  */
-import { Children, createElement } from '@wordpress/element';
+import {
+	createElement,
+	cloneElement,
+	useState,
+	useEffect,
+} from '@wordpress/element';
+import classnames from 'classnames';
+import { MediaUpload } from '@wordpress/media-utils';
 
 /**
  * Internal dependencies
  */
-import { Sortable, SortableHandle } from '../sortable';
+import { Sortable, moveIndex } from '../sortable';
+import { ImageGalleryToolbar } from './index';
+import { ImageGalleryChild, MediaUploadComponentType } from './types';
+import { removeItem, replaceItem } from './utils';
 
 export type ImageGalleryProps = {
-	children: JSX.Element | JSX.Element[];
+	children: ImageGalleryChild | ImageGalleryChild[];
 	columns?: number;
+	onRemove?: ( props: {
+		removeIndex: number;
+		removedItem: ImageGalleryChild;
+	} ) => void;
+	onReplace?: ( props: {
+		replaceIndex: number;
+		previousItem: ImageGalleryChild;
+	} ) => void;
+	onOrderChange?: ( items: ImageGalleryChild[] ) => void;
+	MediaUploadComponent?: MediaUploadComponentType;
 } & React.HTMLAttributes< HTMLDivElement >;
 
 export const ImageGallery: React.FC< ImageGalleryProps > = ( {
 	children,
 	columns = 4,
+	onOrderChange = () => null,
+	onRemove = () => null,
+	onReplace = () => null,
+	MediaUploadComponent = MediaUpload,
 }: ImageGalleryProps ) => {
+	const [ activeToolbarKey, setActiveToolbarKey ] = useState< string | null >(
+		null
+	);
+	const [ orderedChildren, setOrderedChildren ] = useState<
+		ImageGalleryChild[]
+	>( [] );
+
+	useEffect( () => {
+		if ( ! children ) {
+			return;
+		}
+		setOrderedChildren(
+			( Array.isArray( children ) ? children : [ children ] ).map(
+				( child, index ) =>
+					cloneElement( child, { key: child.key || String( index ) } )
+			)
+		);
+	}, [ children ] );
+
+	const updateOrderedChildren = ( items: ImageGalleryChild[] ) => {
+		setOrderedChildren( items );
+		onOrderChange( orderedChildren );
+	};
+
 	return (
 		<div
 			className="woocommerce-image-gallery"
 			style={ {
 				gridTemplateColumns: 'min-content '.repeat( columns ),
 			} }
 		>
-			<Sortable isHorizontal>
-				{ Children.map( children, ( child ) => (
-					<SortableHandle>{ child }</SortableHandle>
-				) ) }
+			<Sortable
+				isHorizontal
+				onOrderChange={ ( items ) => {
+					updateOrderedChildren( items );
+				} }
+			>
+				{ orderedChildren.map( ( child, childIndex ) => {
+					const isToolbarVisible = child.key === activeToolbarKey;
+					const isCoverItem = childIndex === 0;
+
+					return cloneElement(
+						child,
+						{
+							isCover: isCoverItem,
+							className: classnames( {
+								'is-showing-toolbar': isToolbarVisible,
```

Comment:
Nitpicky, but would be nice if these names matched:
```suggestion
								'is-toolbar-visible': isToolbarVisible,
```